terraso_backend/apps/collaboration/models/memberships.py

```diff
@@ -0,0 +1,165 @@
+        user = User.objects.get(email=user_email)
+
+        membership = self.get_membership(user)
+        is_new = not membership
+
+        previous_membership_status = membership.membership_status if not is_new else None
+
+        if validation_func(
+            {
+                "user_role": user_role,
+                "membership_status": membership_status,
+                "current_membership": membership,
+            }
+        ):
+            raise ValidationError("User cannot request membership")
+
+        if is_new:
+            membership = Membership(
+                membership_list=self,
+                user=user,
+                user_role=user_role,
+                membership_status=membership_status,
+            )
+        else:
+            membership.user_role = user_role
+            if membership_status:
+                membership.membership_status = membership_status
+
+        membership_status = membership.membership_status
+
+        is_membership_approved = (
+            previous_membership_status is not None
+            and previous_membership_status != Membership.APPROVED
+            and membership_status == Membership.APPROVED
+        )
+
+        membership.save()
+        return is_membership_approved, membership
+
+    @property
+    def approved_members(self):
+        member_memberships = models.Subquery(self.memberships.approved_only().values("user_id"))
+        return User.objects.filter(id__in=member_memberships)
+
+    def has_role(self, user: User, role: str) -> bool:
+        return self.memberships.by_role(role).filter(id=user.id).exists()
+
+    def is_member(self, user: User) -> bool:
+        return self.approved_members.filter(id=user.id).exists()
+
+    def get_membership(self, user: User):
+        return self.memberships.filter(user=user).first()
+
+    @property
+    def can_join(self):
+        return self.enroll_method in (self.ENROLL_METHOD_JOIN, self.ENROLL_METHOD_BOTH)
+
+    @classmethod
+    def get_membership_type_from_text(cls, membership_type):
+        if membership_type and membership_type.lower() == cls.MEMBERSHIP_TYPE_CLOSED:
+            return cls.MEMBERSHIP_TYPE_CLOSED
+        return cls.MEMBERSHIP_TYPE_OPEN
+
+
+class MembershipObjectsManager(SafeDeleteManager):
+    def by_role(self, role):
+        return self.filter(user_role=role, membership_status=Membership.APPROVED)
+
+    def approved_only(self):
+        return self.filter(membership_status=Membership.APPROVED)
+
+
+class Membership(BaseModel):
+    APPROVED = "approved"
+    PENDING = "pending"
+
+    APPROVAL_STATUS = (
+        (APPROVED, _("Approved")),
+        (PENDING, _("Pending")),
+    )
+
+    membership_list = models.ForeignKey(
+        MembershipList, on_delete=models.CASCADE, related_name="memberships"
+    )
+    user = models.ForeignKey(
+        User, on_delete=models.CASCADE, related_name="collaboration_memberships"
+    )
+
+    user_role = models.CharField(max_length=64)
```

Comment:
Garo also has some comments about this, to avoid duplication I will add my thoughts there, but I agree an improvement will be to restrict the options here somehow, see: https://github.com/techmatters/terraso-backend/pull/718#discussion_r1316486680